dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/dataintegrity/hibernate/HibernateDataIntegrityStore.java

```diff
@@ -0,0 +1,106 @@
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.dataintegrity.hibernate;
+
+import static java.util.stream.Collectors.toUnmodifiableList;
+
+import java.util.Date;
+import java.util.List;
+
+import lombok.AllArgsConstructor;
+
+import org.hibernate.SessionFactory;
+import org.hisp.dhis.dataintegrity.DataIntegrityCheck;
+import org.hisp.dhis.dataintegrity.DataIntegrityDetails;
+import org.hisp.dhis.dataintegrity.DataIntegrityDetails.DataIntegrityIssue;
+import org.hisp.dhis.dataintegrity.DataIntegrityStore;
+import org.hisp.dhis.dataintegrity.DataIntegritySummary;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * As we want each check to be its own transaction the @{@link Transactional}
+ * annotation is used on the store and not the service level in this case.
+ *
+ * @author Jan Bernitt
+ */
+@Repository
+@AllArgsConstructor
+public class HibernateDataIntegrityStore implements DataIntegrityStore
+{
+
+    private final SessionFactory sessionFactory;
+
+    @Override
+    @Transactional( readOnly = true )
+    public DataIntegritySummary querySummary( DataIntegrityCheck check, String sql )
+    {
+        Object[] summary = (Object[]) sessionFactory.getCurrentSession()
+            .createNativeQuery( sql ).getSingleResult();
+        return new DataIntegritySummary( check, new Date(), null, parseCount( summary[0] ),
+            parsePercentage( summary[1] ) );
+    }
+
+    @Override
+    @Transactional( readOnly = true )
+    public DataIntegrityDetails queryDetails( DataIntegrityCheck check, String sql )
+    {
+        @SuppressWarnings( "unchecked" )
+        List<Object[]> rows = sessionFactory.getCurrentSession().createNativeQuery( sql ).list();
+        return new DataIntegrityDetails( check, new Date(), null, rows.stream()
+            .map( row -> new DataIntegrityIssue( (String) row[0],
```

Comment:
Maybe a little dense/hard to read? Maybe put the forth param to DataIntegrityDetails on the line above/

Reply:
Had to merge this PR into #9814 - will try making it more readable there once this is merged so I can resolve the conflicts with master.

Reply:
I introduced a helper method here https://github.com/dhis2/dhis2-core/pull/9814/files#diff-27419f335112488a229aa6db65ac231250233db214802e8e4d951725f565c5d8R79-R81 to make this more readable.